### PR TITLE
Add Sentinel SOFARPC adapter module

### LINKED ISSUE #1189: Support SOFARPC
是否支持 sofarpc 呀？

### PR DESCRIPTION
**Pull request by cdfive**
### Describe what this PR does / why we need it

Add Sentinel adapter to support SOFARPC.

### Does this pull request fix one issue?

Fixes #1189

### Describe how you did it

Learn from SOFARPC official document and also refer to Dubbo adapter.

* Add a adapter module `sentinel-sofa-rpc-adapter`, add custom filter`SentinelSofaRpcProviderFilter` and `SentinelSofaRpcConsumerFilter` to add sentinel
support.

* Add a demo module `sentinel-sofa-rpc-demo` to verify it.

### Describe how to verify it

Run test case and demo.

### Special notes for reviews

* The resourceName of service method does a small improvement, `interfaceName:methodSignature`changed to`interfaceName#methodSignature`, since it's convenient for locate the class by copying it into IDEA with double click shift.

* Now it only support sync invoke,  I'm learning and will try it later.

### CODE REVIEW

File: sentinel-adapter/sentinel-sofa-rpc-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/sofa/rpc/SentinelSofaRpcProviderFilter.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        try {
+            ContextUtil.enter(methodResourceName, applicationName);
+
+            interfaceEntry = SphU.entry(interfaceResourceName, ResourceTypeConstants.COMMON_RPC, EntryType.IN);
+            methodEntry = SphU.entry(methodResourceName, ResourceTypeConstants.COMMON_RPC, EntryType.IN, getMethodArguments(request));
+
+            SofaResponse response = invoker.invoke(request);
+
+            Object appResponse = response.getAppResponse();
+            if (appResponse instanceof Throwable) {
```
Comment (sczyh30) [inline sentinel-adapter/sentinel-sofa-rpc-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/sofa/rpc/SentinelSofaRpcProviderFilter.java]: According to the comments of `SofaResponse`, the `appResponse` may indicate the biz exception and the `errorMsg` indicates the framework exception. Should we handle the latter case?

Comment (cdfive) [reply sentinel-adapter/sentinel-sofa-rpc-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/sofa/rpc/SentinelSofaRpcProviderFilter.java]: Oh, I always thought biz exception should be traced, as it often happened in actual project.
I found the commment in `Trace` class, `Tracer` is used to record other exceptions except block exception. Should we handle both of them?

Besides, I have a question why`com.alibaba.csp.sentinel.adapter.dubbo.SentinelDubboProviderFilter` doesn't catch other Exception exception `BlockException` and `RpcException`?

Comment (sczyh30) [reply sentinel-adapter/sentinel-sofa-rpc-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/sofa/rpc/SentinelSofaRpcProviderFilter.java]: > I found the comment in Trace class, Tracer is used to record other exceptions except for block exception. Should we handle both of them?

Yes, though framework error does not occur frequently, we may need to handle it.

File: sentinel-adapter/sentinel-sofa-rpc-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/sofa/rpc/SentinelSofaRpcProviderFilter.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        Entry methodEntry = null;
+        try {
+            ContextUtil.enter(methodResourceName, applicationName);
+
+            interfaceEntry = SphU.entry(interfaceResourceName, ResourceTypeConstants.COMMON_RPC, EntryType.IN);
+            methodEntry = SphU.entry(methodResourceName, ResourceTypeConstants.COMMON_RPC, EntryType.IN, getMethodArguments(request));
+
+            SofaResponse response = invoker.invoke(request);
+
+            response.setErrorMsg("test error msg");
```
Comment (sczyh30) [inline sentinel-adapter/sentinel-sofa-rpc-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/sofa/rpc/SentinelSofaRpcProviderFilter.java]: Test messages? :)

File: sentinel-adapter/sentinel-sofa-rpc-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/sofa/rpc/SentinelSofaRpcProviderFilter.java [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                SofaRpcException rpcException = new SofaRpcException(RpcErrorType.SERVER_FILTER, response.getErrorMsg());
+                Tracer.traceEntry(rpcException, interfaceEntry);
+                Tracer.traceEntry(rpcException, methodEntry);
+            } else {
+                Object appResponse = response.getAppResponse();
+                if (appResponse instanceof Throwable) {
+                    Tracer.traceEntry((Throwable) appResponse, interfaceEntry);
+                    Tracer.traceEntry((Throwable) appResponse, methodEntry);
+                }
+            }
```
Comment (sczyh30) [inline sentinel-adapter/sentinel-sofa-rpc-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/sofa/rpc/SentinelSofaRpcProviderFilter.java]: Maybe this code snippet can be extracted into `AbstractSofaRpcFilter`?

File: sentinel-adapter/sentinel-sofa-rpc-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/sofa/rpc/SentinelSofaRpcProviderFilter.java (hunk 2 above)
Comment (cdfive) [reply sentinel-adapter/sentinel-sofa-rpc-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/sofa/rpc/SentinelSofaRpcProviderFilter.java]: Sorry for commit test mesage.It has been fixed.

File: sentinel-adapter/sentinel-sofa-rpc-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/sofa/rpc/SentinelSofaRpcProviderFilter.java (hunk 1 above)
Comment (cdfive) [reply sentinel-adapter/sentinel-sofa-rpc-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/sofa/rpc/SentinelSofaRpcProviderFilter.java]: Got it, thanks.

File: sentinel-adapter/sentinel-sofa-rpc-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/sofa/rpc/SentinelSofaRpcProviderFilter.java (hunk 3 above)
Comment (cdfive) [reply sentinel-adapter/sentinel-sofa-rpc-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/sofa/rpc/SentinelSofaRpcProviderFilter.java]: Good idea! Two method, one for trace response exception, and one for trace other exception has been refactored into `AbstractSofaRpcFilter`.

**Comment by sczyh30** [comment]
👍 Excellent work. Thanks!

**Pull request merged**